src/server/pkg/work/testing/work_test.go

```diff
@@ -0,0 +1,151 @@
+package testing
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"strconv"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/pachyderm/pachyderm/src/client/pkg/require"
+	"github.com/pachyderm/pachyderm/src/server/pkg/testutil"
+	"github.com/pachyderm/pachyderm/src/server/pkg/work"
+	"golang.org/x/sync/errgroup"
+)
+
+var (
+	subtaskFailure = fmt.Errorf("subtask failure")
+)
+
+func seedStr(seed int64) string {
+	return fmt.Sprint("seed: ", strconv.FormatInt(seed, 10))
+}
+
+func generateRandomString(n int) string {
+	rand.Seed(time.Now().UnixNano())
+	b := make([]byte, n)
+	for i := range b {
+		b[i] = byte('a' + rand.Intn(26))
+	}
+	return string(b)
+}
+
+func test(t *testing.T, workerFailProb, taskCancelProb, subtaskFailProb float64) {
+	seed := time.Now().UTC().UnixNano()
+	rand.Seed(seed)
+	msg := seedStr(seed)
+	require.NoError(t, testutil.WithEtcdEnv(func(env *testutil.EtcdEnv) error {
+		etcdPrefix := generateRandomString(32)
+		numTasks := 10
+		numSubtasks := 10
+		numWorkers := 5
+		// Setup maps for checking creation, processing, and collection.
+		taskMapsFunc := func() []map[string]bool {
+			var taskMaps []map[string]bool
+			for i := 0; i < numTasks; i++ {
+				taskMaps = append(taskMaps, make(map[string]bool))
+			}
+			return taskMaps
+		}
+		subtasksCreated := taskMapsFunc()
+		subtasksProcessed := taskMapsFunc()
+		var processMu sync.Mutex
+		subtasksCollected := taskMapsFunc()
+		// Setup workers.
+		for i := 0; i < numWorkers; i++ {
+			go func() {
+				for {
+					ctx, cancel := context.WithCancel(context.Background())
+					w := work.NewWorker(context.Background(), env.EtcdClient, etcdPrefix, "")
```

Comment:
It appears that this is necessary for the test to pass (`TestWorkerCrashes` hangs otherwise).  I'm not sure why, but I would guess that canceling the context leaves the worker in a bad state.

Reply:
This is intended to simulate the worker machine crashing. Calling run multiple times might be an issue with the queuing (since it is setup in the `NewWorker` call). I might need to make some changes here.